## Лабораторная работа "Введение в ML"

В этой лабораторной вы:

- познакомитесь с базовыми библиотеками для работы с табличными данными — `numpy` и `pandas`
- поближе посмотрите на простейшие задачи машинного обучения: классификацию и регрессию
- попробуете несколько метрик и поймёте, почему выбор метрики это важно
- обучите несколько простых моделей
- увидите связь между сложностью модели и переобучением
- убедитесь, что без данных всё тлен

Загрузка самых базовых библиотек:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from sklearn.model_selection import train_test_split

### [NumPy](https://numpy.org/doc/stable/user/index.html)

С 1995 numeric, с 2006 NumPy — «Numerical Python extensions» или просто «NumPy»

Возможности библиотеки NumPy:
* работать с многомерными массивами (таблицами)
* быстро вычислять математические функций на многомерных массивах

Ядро пакета NumPy — объект [ndarray](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html)

**Важные отличия** между NumPy arrays и Python sequences:
* NumPy array имеет фиксированную длину, которая определяется в момент его создания (в отличие от Python lists, которые могут расти динамически)
* Элементы в NumPy array должны быть одного типа
* Можно выполнять операции непосредственно над NumPy arrays

**Скорость** NumPy достигается с помощью:
* реализации на C
* векторизации и броадкастинга (broadcasting). Например, произведение массивов совместимых форм.

Теперь давайте разберёмся подробнее и сделаем что-нибудь приятное и полезное в `numpy`!

### Индексация

В NumPy работает привычная индексация Python, ура! Включая использование отрицательных индексов и срезов (slices)

<div class="alert alert-info">
<b>Замечание 1:</b> Индексы и срезы в многомерных массивах не нужно разделять квадратными скобками,
т.е. вместо <b>matrix[i][j]</b> нужно использовать <b>matrix[i, j]</b>. Первое тоже работает, но сначала выдаёт строку i, потом элемент j в ней.
</div>

<div class="alert alert-danger">
<b>Замечание 2:</b> Срезы в NumPy создают view, а не копии, как в случае срезов встроенных последовательностей Python (string, tuple and list).
</div>

In [2]:
ones_matrix = np.ones((5, 5))
ones_submatrix_view = ones_matrix[::2,::2] # creates a view, not copy
ones_matrix[::2,::2] = np.zeros((3, 3))
ones_submatrix_view

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

### Ссылка на Яндекс.Контест

Решения и ответы в задачах, расположенных ниже, загружайте в контест на автоматическую проверку:
https://new.contest.yandex.ru/60376/start


**1.** Реализуйте функцию, принимающую на вход два одномерных массива `first_array` и `second_array` и возвращающую матрицу, в которой первый массив соответствует первому столбцу матрицы, второй — второму.

Вероятно первое, что приходит вам на ум, это конкатенация и транспонирование:

In [3]:
def construct_matrix(first_array, second_array):
    """
    Construct matrix from pair of arrays
    :param first_array: first array
    :param second_array: second array
    :return: constructed matrix
    """
    return np.column_stack((first_array, second_array))

In [4]:
construct_matrix(np.array([1,2,3]),np.array([3,4, 5]))

array([[1, 3],
       [2, 4],
       [3, 5]])

(в скобках заметим, что конкатенировать можно vertically, horizontally, depth wise методами vstack, hstack, dstack по трём осям (0, 1 и 2, соотвественно), либо в общем случае `np.concatenate` — поиграйтесь ниже с прекрасным примером четырёхмерной точки, чтобы точно всё для себя понять)

In [5]:
p = np.arange(1).reshape([1, 1, 1, 1])
p

array([[[[0]]]])

In [6]:
print("vstack: ", np.vstack((p, p)).shape)
print("hstack: ", np.hstack((p, p)).shape)
print("dstack: ", np.dstack((p, p)).shape)

vstack:  (2, 1, 1, 1)
hstack:  (1, 2, 1, 1)
dstack:  (1, 1, 2, 1)


In [7]:
np.concatenate((p, p), axis=3).shape

(1, 1, 1, 2)

Но, поскольку операция транспонирования [делает массив non-contiguous](https://numpy.org/doc/stable/user/basics.copies.html#other-operations), мы в этой задаче **запретим** ей пользоваться и порекомедуем воспользоваться, например, методом [reshape](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html).

**2.** Реализуйте функцию, принимающую на вход массив целых неотрицательных чисел `nums` и возвращающую самый частый элемент массива.

In [8]:
def most_frequent(nums):
    """
    Find the most frequent value in an array
    :param nums: array of ints
    :return: the most frequent value
    """
    unique, counts = np.unique(nums, return_counts=True)
    return unique[np.argmax(counts)]

### Переходим к работе с данными

Прежде всего, загрузим данные и сделаем из них красивые pandas-таблички. Они взяты из параллели RecSys соревнования https://yandex.ru/cup/ml/. Но мы будем иметь дело не со всеми данными, а только с их частью. Данные у нас будут про заведения общественного питания (больше бюрократический терминологии!)

Файлы с данными можно найти [здесь](https://disk.yandex.ru/d/YWvCNRQMb7QSQA).

Задачей будет **предсказание среднего чека** (average_bill) по некоторым другим свойствам заведения.

In [9]:
base = 'data/'

In [10]:
data = pd.read_csv(base + 'organisations.csv')
features = pd.read_csv(base + 'features.csv')
rubrics = pd.read_csv(base + 'rubrics.csv')

В основном мы будем работать с табличкой `data`; остальное вам может пригодиться, если вы захотите знать, какое содержание стоит за кодами признаков.

## Изучение данных

Посмотрите на данные. В этом вам поможет метод ``head`` pandas-таблички.

In [11]:
# <Your code here>
data

org_id city  average_bill    rating   rubrics_id  \
0      15903868628669802651  msk        1500.0  4.270968  30776 30774   
1      16076540698036998306  msk         500.0  4.375000        30771   
2       8129364761615040323  msk         500.0  4.000000        31495   
3      15262729117594253452  msk         500.0  4.538813  30776 30770   
4      13418544315327784420  msk         500.0  4.409091        31495   
...                     ...  ...           ...       ...          ...   
68334   4379286080707082909  msk           NaN  3.812500        30774   
68335   7916477189329738565  msk           NaN  4.894231        30776   
68336  12358902585434046825  msk           NaN  4.156250        30774   
68337   1712093598996183140  spb           NaN       NaN  30771 30774   
68338    125079240973229511  spb           NaN  4.633333        30776   

                                                                                                                                                                                                                                                                                                                                             features_id  
0      3501685156 3501779478 20422 3502045016 3502045032 3501618484 11177 11012 2020795524 10894 3501599541 12048 10462 1524 1509 3501498187 11867 1416 3501744275 3501942428 273469383 11704 1189498238 20424 21247 3501481353 3491142672 3501481355 1082283206 11741 3501471642 3501498186 11629 11617 10886 1415 3501513153 1018 246 3501749289 11531  
1                                                                                                                                                                                                                                                                 1509 1082283206 273469383 10462 11617 3501779478 3501744275 1524 1018 1116173086 20422  
2                                                                                                                                                                                                                                                         10462 11177 11617 11629 1416 1018 11704 11867 3501744275 20282 3502045016 273469383 3502045032  
3                                                                                                                                                                                                                                                               3501618484 2020795524 11629 11617 1018 11704 21247 11867 246 20422 1416 3501749289 20305  
4                                                                                                                                                                                                                                                         11617 10462 11177 1416 11867 3501744275 20282 3502045016 11629 273469383 3502045032 1018 11704  
...                                                                                                                                                                                                                                                                                                                                                  ...  
68334                                                                                                                1018 1415 10462 11629 11867 20422 20424 1189498238 3491142672 3501481353 11617 11704 3501637468 3501744275 3501773763 246 3501779478 12048 3501773764 3501942428 3502010918 3502010929 21247 3502010932 11741 3501513153 3502010938  
68335                                                                                                                          11634 11629 3501481353 11177 3501773763 11867 20422 2020795524 11617 1018 3501745827 246 3501779478 3501754799 11741 3501513153 21247 137 3501643040 273469383 1415 1416 3501945049 1509 1524 3501773764 12048 3501942428  
68336                

In [12]:
features.head()

feature_id                           feature_name
0           1  prepress_and_post_printing_processing
1          40                               products
2          54                        printing_method
3          77                                   fuel
4          79                                   shop

In [13]:
rubrics.head()

rubric_id        rubric_name
0      30519  Булочная, пекарня
1      30770           Бар, паб
2      30771    Быстрое питание
3      30774               Кафе
4      30775           Пиццерия

Полезно посмотреть внимательнее на то, с какими признаками нам предстоит работать.

* **org_id** вам не понадобится;
* **city** - город, в котором находится заведение (``msk`` или ``spb``);
* **average_bill** - средний чек в заведении - он будет нашим таргетом;
* **rating** - рейтинг заведения;
* **rubrics_id** - тип заведения (или несколько типов). Соответствие кодов каким-то человекочитаемым типам живёт в табличке ``rubrics``
* **features_id** - набор неких фичей заведения. Соответствие кодов каким-то человекочитаемым типам живёт в табличке ``features``

Обратите внимание, что **rubrics_id** и **features_id** - это не списки, а разделённые пробелами строки. Когда вам захочется работать с отдельными фичами из мешка фичей для данного заведения, вам придётся всё-таки превратить их в списки (здесь поможет метод `split` для строк).

Чтобы быстро восстанавливать по рубрикам и фичам их нормальные названия, сделайте словари вида ``код_фичи:название_фичи``

In [14]:
# <Your code here>
feature_dict = features.set_index("feature_id").to_dict(orient='dict')
rubric_dict = rubrics.set_index("rubric_id").to_dict(orient='dict')
rubric_dict

{'rubric_name': {30519: 'Булочная, пекарня',
  30770: 'Бар, паб',
  30771: 'Быстрое питание',
  30774: 'Кафе',
  30775: 'Пиццерия',
  30776: 'Ресторан',
  30777: 'Столовая',
  31286: 'Спортбар',
  31350: 'Кондитерская',
  31375: 'Суши-бар',
  31401: 'Кальян-бар',
  31495: 'Кофейня',
  3108292683: 'Бар безалкогольных напитков',
  3501514558: 'Фудкорт',
  3501750896: 'Кофе с собой'}}

Посмотрим, какими бывают типы заведений:

In [15]:
rubric_dict

{'rubric_name': {30519: 'Булочная, пекарня',
  30770: 'Бар, паб',
  30771: 'Быстрое питание',
  30774: 'Кафе',
  30775: 'Пиццерия',
  30776: 'Ресторан',
  30777: 'Столовая',
  31286: 'Спортбар',
  31350: 'Кондитерская',
  31375: 'Суши-бар',
  31401: 'Кальян-бар',
  31495: 'Кофейня',
  3108292683: 'Бар безалкогольных напитков',
  3501514558: 'Фудкорт',
  3501750896: 'Кофе с собой'}}

Мы что-то поняли про признаки, которыми нам предстоит пользоваться. Теперь время посмотреть на таргет. Вооружившись функциями ``hist`` и ``scatter`` из библиотеки ``matplotlib``, а также методом ``isna`` для pandas-таблиц разберитесь, какие значения принимают таргеты, есть ли там там выбросы, пропуски или ещё какие-то проблемы.

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    <ol>
      <li>Среди таргетов довольно много пропусков;</li>
      <li>Все таргеты - это числа, кратные 500;</li>
      <li>Есть какие-то адские значения, превышающие 100 000 (видимо, выбросы);</li>
      <li>В целом, число ресторанов с данным средним чеком быстро падает с ростом среднего чека. Для средних чеков, больших 2500, заведений уже совсем мало. Примерно у 2/3 заведений средний чек 500.</li>
    </ol>
</details>

In [16]:
data.isna().sum()

org_id              0
city                0
average_bill    35561
rating          13731
rubrics_id          0
features_id     11049
dtype: int64

**Базовая очистка данных**

Раз есть треш, давайте чистить данные.

С пропусками можно бороться по-разному (даже и с пропусками в таргете), но пока мы сделаем самую простую вещь: дропнем все заведения, для которых мы не знаем средний чек.

Уберите из них все заведения, у которых средний чек неизвестен или превышает 2500. Пока есть опасение, что их слишком мало, чтобы мы смогли обучить на них что-нибудь.

**3. Введите в Контест количество заведений, которое у вас получилось после очистки**.

Дальше мы будем работать с очищенными данными.

In [17]:
# <Your code here>
data_filt = data.dropna(subset=['average_bill'])
data_filt = data_filt[data_filt["average_bill"] <= 2500]
data_filt.shape

(32136, 6)

**4. Посчитайте и введите в Контест разность между средними арифметическими average_bill в кафе Москвы и Санкт-Петербурга. Округлите ответ до целого.**

&nbsp;

<details>
  <summary>Небольшая подсказка</summary>
  Примените часто используемый метод groupby.
</details>

In [18]:
data_filt.groupby("city")["average_bill"].mean()

city
msk    792.887230
spb    676.449662
Name: average_bill, dtype: float64

In [19]:
msk_avg = data_filt[data_filt['city'] == 'msk']['average_bill'].mean()

# Считаем среднее значение average_bill для Санкт-Петербурга
spb_avg = data_filt[data_filt['city'] == 'spb']['average_bill'].mean()

# Разница между средними значениями
difference = round(abs(msk_avg - spb_avg))
difference

116

Давайте ещё немного поизучаем данные. Ответьте на вопросы:

1. Есть ли разница между средними чеками в Москве и Санкт-Петербурге?
2. Коррелирует ли средний чек с рейтингом?
3. Есть ли разница в среднем чеке между ресторанами и пабами (см. соответствующие типы из ``rubrics``)?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    <ol>
      <li>В целом, да. Вы могли бы сравнить средние (в Москве больше) или медианы (они равны, потому что уж больно много где средний чек 500). Этого, конечно, мало для того, чтобы сделать вывод. Нужно проверять какие-то статические критерии, которые изучаются в курсе по статистике. Не будем останавливаться на этом подробно. Поскольку данные совсем не нормальные, никакой t-тест не сработает; мы бы предложили использовать критерий Манна-Уитни (см. википедию и функцию mannwhitneyu из библиотеки scipy.stats).</li>
      <li>Какая-то корреляция между ними есть но уж больно неубедительная (рекомендуем построим на одном графике boxplot рейтинга по каждому значению среднего чека для визуализации). Конечно, дна становится меньше с ростом среднего чека, но, видимо, в предсказании это особо не используешь;</li>
      <li>Несомненно, в ресторанах средний чек выше. Это и невооружённым глазом видно, и с помощью критерия Манна-Уитни можно проверить.</li>
    </ol>
</details>

## Формулируем задачу

Прежде, чем решать задачу, её надо сформулировать.

**Вопрос первый**: это классификация или регрессия? Подумайте над этим.

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    Ответ не столь однозначен, как хотелось бы. С одной стороны, таргет принимает всего четыре значения, и потому это может быть классификацией с 4 классами. С другой стороны, таргеты - это не абстрактные "треугольник", "круг", "квадрат", а вещественные числа, и когда мы вместо 500 предсказываем 2500, это явно хуже, чем вместо 1500 предсказать 2000. В целом, задачу можно решать и так, и так; мы будем смотреть на метрики обеих задач.
</details>

**Вопрос второй**: какие метрики мы будем использовать для оценки качества решения? Какие метрики вы предложили бы для этой задачи как для задачи классификации? А для этой задачи, как для задачи регрессии?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    
    Начнём с классификации. Метрика accuracy не очень хороша из-за несбалансированности классов. Действительно, классификатор, который всегда говорит 500, будет иметь accuracy примерно 0.66, хотя это никак не отражает практическую ценность модели. Как мы увидим, самая большая проблема будет заключаться в том, чтобы научиться выделять заведения с большими чеками, а их меньше всего и в accuracy они вносят самый маленький вклад. Есть разные способы с этим бороться, один -- использовать sklearn.metrics.balanced_accuracy_score. Его идея, грубо говоря, в том, чтобы по каждому классу найти, какая доля объектов этого класса правильно классифицирована, а потом эти доли усреднить. Тогда у бессмысленного классификатора, который всем ставит 500, будет скор 1/5 (ведь классов 5), а чтобы получить прежние 2/3, нужно будет научиться в каждом классе правильно ставить хотя бы 2/3 меток.    
    
    Теперь что касается регрессии. Основых метрики две - MSE и MAE. Из первой стоит извлекать корень, чтобы получать интерпретируемые человеком значения, а вторая менее агрессивна к выбросам (впрочем, выбросов тут уже нет, мы их все выкинули). Без дополнительной информации не очень понятно, какую выбирать, можно брать любую. А выбирать надо: ведь даже банальные модели "предсказывай всегда среднее" и "предсказывай всегда медиану" будут по-разному ранжироваться этими метриками.
    
</details>

**Вопрос третий**: а не взять ли нам какую-нибудь более экзотическую метрику? Например, MAPE (определение в учебнике в главе про оценку качества моделей). А как вам такое соображение: допустим, заказчик говорит, что пользователи будут расстраиваться, только если мы завысили средний чек - так давайте поправим MSE или MAE, обнуляя те слагаемые, для которых предсказанный таргет меньше истинного. Вот это хорошая метрика или нет?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    
    Что касается MAPE, у нас нет тех проблем, с которой она борется. Вот если бы у нас были средние чеки от 500 до миллиона, мы бы столкнулись с ситуацией, что большие ошибки для больших чеков доминировали бы в сумме для MSE и MAE (500 вместо 1000 меркнет по сравнению с 500к вместо миллиона). Говоря поэтически, мы бы оптимизировали модель для миллионеров, забыв про простых трудяг. И было бы логично перейти от парадигмы "ошибаемся на 500 рублей" к парадигме "ошибаемся на 50%". Но у нас все таргеты примерно одного порядка, MAPE нам особо ни к чему.
    
    Вторая метрика коварна тем, что её можно "накрутить" безо всякой пользы для дела. А именно, модель, которая всегда предсказывает средний чек в миллион, была бы идеальна. Но все бы расстраивались и не ходили есть. Другое дело, что можно ввести разные веса для ошибок в большую и в меньшую сторону, но опять же - пока нет показаний к тому, что это нужно.
    
</details>

## Применяем ML

Теперь время разбить данные на обучающую и тестовую выборку. Делается это с помощью функции ``train_test_split`` из пакета ``sklearn``. При этом очень важно сделать две вещи:

* Зафиксировать ``random_state=42`` (да, именно этот, а то ваши модели могут не зайти в Контест), чтобы всё, что мы делаем, было воспроизводимо (иначе от перезапуска к перезапуску числа могут меняться, и мы не будем понимать, из-за чего это происходит).
* Сделать стратификацию по таргету. В противном случае у нас в трейне и тесте могут оказаться разные пропорции классов (обычно особенно страдают мало представленные классы), что неутешительно скажется на результате.

**Обратите внимание**, что если вы побьёте выборку на train и test по-другому, ваши результаты могут не зайти в контест.

In [20]:
clean_data = data_filt
clean_data_train, clean_data_test = train_test_split(
    clean_data, stratify=clean_data["average_bill"], test_size=0.33, random_state=42)

Теперь нам нужен **бейзлайн** - очень простая модель, с которой мы в дальнейшем будем сравниваться.

Поскольку мы ещё не знаем никаких умных классов моделей, все модели мы будем писать руками. А именно, мы напишем две простых модели на основе ``sklearn.baseRegressorMixin`` и ``sklearn.base.ClassifierMixin`` (посмотрите примеры в документации sklearn и сделайте так же):

* Модель для задачи регрессии, которая для всех заведений предсказывает одно число — среднее значение среднего чека;
* Модель для задачи классификации, которая для всех заведений предсказывает один класс — самый частый класс (ироничным образом он в данном случае совпадает с медианой).

**Важно!** Мы будем много раз повторять вам мантру о том, что **информация из тестовой выборки не должна протекать в процесс обучения**. Так вот, и среднее, и самый частый класс вы должны считать именно на обучающей выборке!

**5 и 6. Напишите эти две модели и сдайте в Контест**. В процессе проверки модели будут и обучаться, и предсказывать.

Заметим, что для этих моделей нам вообще не нужны какие-то "фичи"; мы работаем только с таргетом.

У каждой модели есть (как минимум) два метода: `fit` (обучает модель по фичам `X` и таргету `y`) `predict` (предсказывает по фичам `X`)

In [21]:
clean_data_train

org_id city  average_bill    rating   rubrics_id  \
45769   3276960721840719260  msk         500.0  4.500000        30770   
39061   8452997364765928283  msk        1500.0  4.442623  30774 30776   
59281  14240408259222214074  spb        1000.0  4.018868  30776 30774   
51225  15114069072602161053  msk        1500.0  4.364742  31401 30776   
29587   2730337118800634815  msk        1000.0  4.698718        30770   
...                     ...  ...           ...       ...          ...   
64667  15641319025413596274  msk         500.0  4.510753        30771   
47309   2049892259403324519  msk         500.0  4.333333        30771   
26208  12224074314753892871  msk         500.0  5.000000        30775   
48599  16581456988770474074  msk         500.0  4.692308  31495 30774   
17863  11409605273003015299  msk         500.0  3.222222        30774   

                                                                                                                                                                                                                                             features_id  
45769                                                                                                                                                                                                            11704 20422 1018 11177 1416 11867 10462  
39061                                                                                                                      1415 3501481355 1416 11629 10462 1524 20422 11867 1018 246 20424 11177 21247 1509 11617 273469383 11704 1082283206 3501618484  
59281                                                                   3502045032 11741 3502045016 10462 11704 3501779478 3501773764 1509 11629 3501744275 11617 1018 11177 11867 12048 246 20422 3501618484 1416 1082283206 20424 3501773763 273469383  
51225                                                                              3501513153 3501779478 3491142672 273469383 3501481355 21247 20424 20422 11741 11629 3501744275 11617 11867 11177 11704 10462 1415 1018 3501481353 246 1082283206 1416  
29587  21247 10896 3491142672 11629 3501481353 3501481355 10428 3501773763 2020795524 3501685156 3501773764 10900 3501618484 1524 20305 3501471642 1018 10894 10462 1416 10888 10770 11867 10875 246 3501779478 11709 11704 11617 1082283206 20422 10886  
...                                                                                                                                                                                                                                                  ...  
64667                                                                                                                                                                     20424 3501744275 273469383 10462 11177 11617 11629 1018 11704 11867 1416 20422  
47309                                                                                                                                                                                                       273469383 20424 11704 11629 10462 20422 1018  
26208                                                                                                                            21247 11867 11629 1524 1509 20422 1416 1415 1018 11704 246 1082283206 1189498238 3501481353 3501513153 10462 3501522239  
48599                                                                            3491142672 20282 3501637468 11741 3501745827 3501694289 3501779478 3501773764 11704 10462 273469383 3501773763 1018 1415 11867 3501769434 11177 11617 20424 20422 11629  
17863                                                                                                                                         3501744275 11617 3502045032 3502045016 20424 10462 12048 11177 11704 1018 273469383 20422 1416 11867 11629  

[21531 rows x 6 columns]

In [22]:
from scipy.stats import mode
import numpy as np
from sklearn.base import RegressorMixin

class MeanRegressor(RegressorMixin):
    # Predicts the mean of y_train
    def __init__(self, *, param=1):
        self.result = 0
        self.is_fitted = None
        self.param = param
    def fit(self, X=None, y=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Training data features
        y : array like, shape = (_samples,)
        Training data targets
        '''
        self.is_fitted = True
        self.result = np.mean(y)
        return self

    def predict(self, X=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Data to predict
        '''
        return [self.result]*X.shape[0]

from sklearn.base import ClassifierMixin

class MostFrequentClassifier(ClassifierMixin):
    # Predicts the rounded (just in case) median of y_train
    def __init__(self, *, param=1):
        self.is_fitted = None
        self.result = None
        self.param = param
    def fit(self, X=None, y=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Training data features
        y : array like, shape = (_samples,)
        Training data targets
        '''
        self.is_fitted = True
        self.result = mode(y)[0]

        return self

    def predict(self, X=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Data to predict
        '''
        return [self.result]*X.shape[0]

Обучим наши модели

In [23]:
reg = MeanRegressor()
reg.fit(y=clean_data_train['average_bill'])

clf = MostFrequentClassifier()
clf.fit(y=clean_data_train['average_bill'])

In [24]:
from sklearn.metrics import mean_squared_error, balanced_accuracy_score
y_reg_predict = reg.predict(X=clean_data_test['average_bill'])
y_clf_predict = clf.predict(X=clean_data_test['average_bill'])
y_true = clean_data_test['average_bill']
print("Regression:RMSE = ", np.sqrt(mean_squared_error(y_true, y_reg_predict)))
print("Classification:RMSE = ", np.sqrt(mean_squared_error(y_true, y_clf_predict)))
print("Classification:BA = ", balanced_accuracy_score(y_true, y_clf_predict))

Regression:RMSE =  448.7143889551622
Classification:RMSE =  514.7517402382093
Classification:BA =  0.2


Обучите модели и оцените их качество на тестовой выборке. В качестве метрик возьмём RMSE (``np.sqrt`` от ``sklearn.metrics.mean_squared_error``) и ``sklearn.metrics.balanced_accuracy_score``.

Для регрессионной модели имеет смысл считать только RMSE (значения будут не кратны 500, точно мы угадывать не будем никогда), а вот для классификационной можно найти обе метрики. Сделайте это. Какая модель оказалась лучше по RMSE?

<details>
  <summary>Когда будете готовы, кликните сюда</summary>
    
  Казалось бы, регрессор никогда не угадывает, но он в каком-то смысле лучше классификатора - справедливо ли это? Возможно. Несуществующий пользователь модели вряд ли будет задавать вопросы "почему средний чек не кратен 500?" Ну, выдали около 800 - ок, понятно.
    
</details>

## Усложнение модели

Бейзлайны будут нашей отправной точкой. Строя дальнейшие модели, мы будем спрашивать себя: получилось ли лучше бейзлайна? Если нет или если не особо, то в чём смысл усложнения?

Начнём с использования фичи ``city``. Мы уже видели, что в разных городах и средние чеки разные. Легко проверить, что *медиана* средних чеков всё же одна и та же и в Москве, и в Санкт-Петербурге (ох уж этот вездесущий средний чек 500!), поэтому с классификатором мы ничего не сделаем. Но вот регрессор можно попробовать починить.

**7. Напишите регрессор, для каждого заведения предсказывающий среднее значение в том же городе (на обучающей выборке, конечно) и сдайте его в Контест**. Вам может помочь то, что булевы `pandas` и `numpy` столбцы можно умножать на численные — в такой ситуации False работает, как ноль, а True как единица.

In [25]:
import pandas as pd
import numpy as np
from sklearn.base import RegressorMixin

class CityMeanRegressor(RegressorMixin):
    def __init__(self, *, param=1):
        self.result = None
        self.city_mean = None
        self.is_fitted = None
        self.param = param
    def fit(self, X=None, y=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Training data features
        y : array like, shape = (_samples,)
        Training data targets
        '''
        if X is None:
            return self
        if y is None:
            return self
        data = pd.DataFrame({'city': X['city'], 'average_bill': y})
        city_avg = data.groupby('city')['average_bill'].mean()
        self.city_mean = city_avg
        self.result = None
        self.is_fitted = True
        return self

    def city_average_predictor(self, city):
        return self.city_mean.get(city, np.nan)
    def predict(self, X=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Data to predict
        '''
        if X is None:
            return self.result
        self.result = [self.city_mean.get(city, np.nan) for city in X["city"]]
        return self.result

In [26]:
from CityMeanRegressor import CityMeanRegressor
city_reg = CityMeanRegressor()
city_reg.fit(X=clean_data_train, y=clean_data_train['average_bill'])
city_reg.predict(X=clean_data_test)

[677.7104465814781,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 677.7104465814781,
 792.6111908177905,
 792.6111908177905,
 677.7104465814781,
 677.7104465814781,
 677.7104465814781,
 792.6111908177905,
 677.7104465814781,
 677.7104465814781,
 677.7104465814781,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 677.7104465814781,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 677.7104465814781,
 792.6111908177905,
 677.7104465814781,
 792.6111908177905,
 677.7104465814781,
 677.7104465814781,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 677.7104465814781,
 677.7104465814781,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 677.7104465814781,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 792.6111908177905,
 677.7104465814781,


In [27]:
from sklearn.metrics import mean_squared_error, balanced_accuracy_score
y_new_reg_predict = city_reg.predict(X=clean_data_test)
y_reg_predict = reg.predict(X=clean_data_test)
y_clf_predict = clf.predict(X=clean_data_test)
y_true = clean_data_test['average_bill']
print("Regression:RMSE = ", np.sqrt(mean_squared_error(y_true, y_reg_predict)))
print("Classification:RMSE = ", np.sqrt(mean_squared_error(y_true, y_clf_predict)))
print("New Regression:RMSE = ", np.sqrt(mean_squared_error(y_true, y_new_reg_predict)))

Regression:RMSE =  448.7143889551622
Classification:RMSE =  514.7517402382093
New Regression:RMSE =  445.1063281403263


Обучите регрессор и сравните его по метрике RMSE с бейзлайнами. Получилось ли улучшить метрику?

Лучше стало, но, правда, не очень сильно. В этот момент очень важно не просто радовать руководителя приростом в третьем знаке, но и думать о том, что происходит.

Средний средний чек по Москве равен 793, в Санкт-Петербурге - 676, а в целом - 752 рубля. MSE, увы, не поможет вам ответить на вопрос, стало ли лучше пользователю, если вы ему вместо 752 рублей назвали 793. Здесь вскрывается весьма существенный порок MSE в этой задаче. Дело в том, что наш изначальный таргет делит заведения на некоторые "ценовые категории", и различие в средних чеках 500 и 1000 в самом деле существенно. Наверное, мы хотели бы как раз правильно предсказывать ценовые категории. Но MSE не очень помогает нам об этом судить. Дальше мы ещё подумаем, как это исправить.

В любом случае, несмотря на улучшение метрики, мы пока не можем судить, стало ли по жизни лучше от усложнения модели.

Поручинившись немного, возьмём на вооружение другую идею. Давайте использовать типы заведений!

Но с типами есть некоторая проблема: в столбце ``rubrics_id`` не всегда один идентификатор, часто их несколько, и всего комбинаций довольно много. Чтобы не возиться с малочисленными типами, давайте сольём их в один безликий ``other``.

Итак, добавьте в обучающие и тестовые данные столбец ``modified_rubrics``, в котором будет то же, что и в ``rubrics_id``, если соответствующая комбинация рубрик содержит хотя бы 100 заведений из обучающей (!) выборки, и строка ``other`` в противном случае.

Здесь вам поможет контейнер ``Counter`` из библиотеки ``collections``.

In [51]:
from collections import Counter

counter = Counter(clean_data_train["rubrics_id"])
clean_data_train["modified_rubrics"] = [rubric if counter[rubric] >= 100 else "other" for rubric in clean_data_train["rubrics_id"]]
clean_data_test["modified_rubrics"] = [rubric if counter[rubric] >= 100 else "other" for rubric in clean_data_test["rubrics_id"]]

Теперь настало время написать могучий классификатор, который по заведению предсказывает медиану средних чеков среди тех в обучающей выборке, у которых с ним одинаковые `modified_rubrics` и город (вы спросите, почему медиану, а не самый частый -- спишем это на вдохновение; самый частый тоже можно брать - но медиана работает лучше).

**8. Напишите классификатор и сдайте в Контест**.

In [88]:
city = 0
by_city = 1
msk = 0
spb = 1
X=clean_data_train
y=clean_data_train['average_bill']
data = pd.DataFrame({'city': X['city'], "modified_rubrics":X["modified_rubrics"], 'average_bill': y})
print("modified_rubrics" in X)
data_by_city = pd.DataFrame(data.groupby('city'))[by_city]
medians_msk = data_by_city[msk].groupby("modified_rubrics")["average_bill"].median().to_dict()
medians_spb = data_by_city[spb].groupby("modified_rubrics")["average_bill"].median().to_dict()

True


In [111]:
from numpy import median
import pandas as pd
from sklearn.base import ClassifierMixin

class RubricCityMedianClassifier(ClassifierMixin):
    # Predicts the rounded (just in case) median of y_train
    def __init__(self, *, param=1):
        self.dict_pro = None
        self.is_fitted = None
        self.result = None
        self.param = param
    def fit(self, X=None, y=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Training data features
        y : array like, shape = (_samples,)
        Training data targets
        '''
        data = pd.DataFrame({'modified_rubrics': X['modified_rubrics'],
                             'city': X['city'],
                             'average_bill': y})
        city = 0
        by_city = 1
        msk = 0
        spb = 1
        data_by_city = pd.DataFrame(data.groupby('city'))[by_city]
        medians_msk = data_by_city[msk].groupby("modified_rubrics")["average_bill"].median().to_dict()
        medians_spb = data_by_city[spb].groupby("modified_rubrics")["average_bill"].median().to_dict()
        self.dict_pro = {"msk": medians_msk, "spb": medians_spb}
        self.is_fitted = True
        return self

    def predict(self, X=None):
        '''
        Parameters
        ----------
        X : array like, shape = (n_samples, n_features)
        Data to predict
        '''
        self.result = [self.dict_pro[city][rubric] for city, rubric in zip(X["city"], X["modified_rubrics"])]
        return self.result

In [115]:
model = RubricCityMedianClassifier()
model.fit(X=clean_data_train, y=clean_data_train['average_bill'])
y_rub_city_clf_predict = model.predict(X=clean_data_test)

In [116]:
from sklearn.metrics import mean_squared_error, balanced_accuracy_score
y_reg_predict = reg.predict(X=clean_data_test)
y_clf_predict = clf.predict(X=clean_data_test)
y_true = clean_data_test['average_bill']
print("Regression:RMSE = ", np.sqrt(mean_squared_error(y_true, y_reg_predict)))
print("Classification:RMSE = ", np.sqrt(mean_squared_error(y_true, y_clf_predict)))
print("Classification:BA = ", balanced_accuracy_score(y_true, y_clf_predict))
print("New Regression:RMSE = ", np.sqrt(mean_squared_error(y_true, y_new_reg_predict)))
print("New Classification:RMSE = ", np.sqrt(mean_squared_error(y_true, y_rub_city_clf_predict)))
print("New Classification:BA = ", balanced_accuracy_score(y_true, y_rub_city_clf_predict))

Regression:RMSE =  448.7143889551622
Classification:RMSE =  514.7517402382093
Classification:BA =  0.2
New Regression:RMSE =  445.1063281403263
New Classification:RMSE =  393.96675836287915
New Classification:BA =  0.30552511833185647


Сравните обученный классификатор по метрикам RMSE и balanced_accuracy_score с нашими бейзлайнами. Получилось ли улучшить?

Обратите внимание что рост accuracy по сравнению с бейзлайном при этом на порядок меньше:

accuracy_score

Predict most frequent:  0.6947666195190948

Predict by rubric and city:  0.7095709570957096

Для диагностики напечатайте для каждого класса тестовой выборки, сколько в нём объектов и скольким из них наш классификатор приписал правильный класс. Что вы видите?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
    
  Вы, вероятно, видите то, что мы стали однозначно лучше по сравнению с бейзлайном детектировать средний чек 1000 и 1500 (хотя всё равно не очень хорошо + ценой ухудшения качества на среднем чеке 500), а вот чеки 2000 и 2500 нам ну никак не даются.
    
</details>

**Кстати**. А вы понимаете, почему приведённый выше пайплайн классификации был не очень удачным с точки зрения архитектуры? Почему его было бы правильнее воплотить по-другому?

&nbsp;

<details>
  <summary>Когда будете готовы, кликните сюда, чтобы посмотреть ответ</summary>
Собственно говоря, и не было никакого пайплайна. К счастью, у нас была одна обучающая выборка, мы на ней посчитали список рубрик для modified_rubrics и радовались жизни. Но если бы нам надо было переобучать всё на новых данных, пришлось бы помнить, что их надо везде пересчитать (ведь у нас могли появиться новые рубрики с хотя бы 100 представителями). А уж никакую кросс-валидацию (кто знает - тот поймёт) с нашим подходом к делу и вовсе бы не получилось сделать без боли.
    
Поэтому в следующей лабораторной вы научитесь делать честные пайплайны, в которых преобразование данных, генерация фичей и обучение классификатора будут объединены в один понятный процесс, происходящий на этапе fit.
</details>

## Слишком простые и слишком сложные модели

Бейзлайны у нас слишком просты и потому не очень полезны в жизни. Но если сложность модели растёт бесконтрольно, то тоже получается плохо.

Давайте рассмотрим конкретный пример. Создадим классификатор, использующий одновременно `rubrics_id` и `features_id`.

Сделайте следующее:

- для каждого объекта обучающей выборки сконкатенируйте строку `rubrics_id` с разделителем (например, буквой 'q') и содержимым `features_id`. Полученный столбец озаглавьте `modified_features`. Это не самый клёвый способ заиспользовать все фичи, но сейчас пока сойдёт. Причём на сей раз не будем выкидывать мало представленные значения (вся информация важна, не так ли?).
- при этом для тестовой выборке заменяйте на строку `other` все конкатенации, которые не встретились в обучающей выборке.

То есть элементы в этом столбце будут иметь вид `other` или `30776 30774 q 3502045032 11741 3502045016 1046...`.

In [117]:
# my code


Теперь обучите классификатор, который для заведения предсказывает медиану среднего чека по всем объектам тестовой выборки с таким же, как у него, значением `modified_features`, а если такого в обучающей выборке нет, то глобальную медиану среднего чека по всей обучающей выборке.

**9. Загрузите в Контест предсказания этого классификатора на тестовой выборке**

Мы ждём файла **.csv**, у которого в каждой строке будет только одно число - предсказание классификатора.

Возможно, вам будет полезна библиотека ``tqdm``, позволяющая отслеживать в реальном времени, сколько времени уже крутится цикл и сколько итераций ещё осталось. Впрочем, если вы всё написали нормально, то должно работать не очень долго.

Модель, очевидно, очень сложная. Число параметров (различных категорий) в ней сопоставимо с числом объектов в обучающей выборке. А получилось ли хорошо?

Давайте посчитаем RMSE и balanced_accuracy_score на обучающей и на тестовой выборках.

**10. Введите их в Контест**

Налицо переобучение: на трейне метрики отличные, на тесте - вообще никакие

В общем, не гонитесь за чрезмерной сложностью модели..

## ML без данных что компутер без электричества

Возможно, вы смотрите на полученные выше результаты и думаете: вот если бы мы не какие-то убогие медианы предсказывали, а гоняли бы нейросети, то тут-то бы всё и получилось!

Но, увы, совсем даже не всегда от счастья нас отделяет выбор хорошей модели (и стратегии обучения). Если данные не очень, то даже самая крутая модель не сработает. В этой ситуации нужно либо добывать новые фичи каким-то образом, либо собирать новые данные (увеличивать датасет), либо просто бросать задачу.

Давайте посмотрим, что выжмет из наших данных одна из самых мощных моделей для табличных данных - градиентный бустинг на решающих деревьях в исполнении [CatBoost](https://catboost.ai/).

Но прежде, чем сделать fit, нам надо облагородить данные. Несмотря на то, что CatBoost отлично работает с категориальными фичами, мешок признаков из `rubrics_id` или `features_id` может ему оказаться не по зубам. Поэтому мы соберём датасет в пристойную матрицу, создав для каждого типа рубрик и фичей отдельный столбец и записав там единицы для тех объектов, у которых эта рубрика или фича имеет место.

В матрице почти все элементы будут нулями. Такие матрицы считаются **разреженными** и их можно хранить гораздо эффективней, чем просто таблицей. Этим и займёмся)

Есть несколько форматов хранения разреженных матриц (многие из них реализованы в [пакете sparse библиотеки scipy](https://docs.scipy.org/doc/scipy/reference/sparse.html)), и каждый пригоден для чего-то своего.

Создавать разреженную матрицу лучше в [формате COO](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_array.html#scipy.sparse.coo_array). Он предполагает, что разреженная матрица задаётся в виде трёх списков: `row`, `col`, `data`, причём каждая тройка `(row[i], col[i], data[i])` кодирует элемент со значением `data[i]`, стоящий на позиции `(row[i], col[i])`. Считается, что на позициях `(row, col)`, которые ни разу не встретились, стоят нули.

Нетрудно видеть, что заполнять такую матрицу - одно удовольствие, и особенно этому помогает тот факт, что **пара `(row, col)` может встретиться несколько раз** (тогда в итоговой матрице на соответствующей позиции стоит сумма соответствующих `data[i]`). Но, с другой стороны, почти ничего другого с такой матрицей не сделаешь: произвольного доступа к элементам она не предоставляет, умножить её тоже особо ничего не умножишь. Поэтому для дальнейшего использования созданную таким образом матрицу преобразуют в один из более удобных форматов, например, [CSR (compressed sparse row)](https://scipy-lectures.org/advanced/scipy_sparse/csr_matrix.html). Он, к примеру, хорошо подходит для умножения на вектор (потому что матрица хранится по строкам). Не будем разбирать его подробно, но можете почитать по ссылке, если интересно.

Вам нужно будет превратить обучающие и тестовые данные в разреженные матрицы `sparse_data_train` и `sparse_data_test` соответственно, таким образом, что:

- столбец `city` превратится в столбец из единиц и нулей (например, 1 - Москва, 0 - Питер);
- столбец `rating` перекочует в разреженные матрицы без изменений;
- каждый типы рубрик и каждая фича превратятся в отдельный 0-1-принак;

В тестовой выборке будут фичи, которых в обучающей выборке не было. С ними можно по-разному работать, но давайте создадим дополнительную фантомную фичу `feature_other`, в которой будет то, сколько неизвестных по обучающей выборке фичей есть у данного объекта.

In [ ]:
# <Your code here>

Данные готовы, и теперь можно запустить катбуст

In [ ]:
from catboost import CatBoostClassifier

In [ ]:
# <USE IT!>
clf = CatBoostClassifier()
clf.fit(sparse_data_train, clean_data_train['average_bill'])

**11. Пришлите в Контест balanced_accuracy_score на тестовой выборке, округлённый до двух знаков после запятой**. Стало ли сильно лучше от того, что мы воспользовались таким крутым классификатором?